### PR TITLE
Dev Finish timesheet api

### PR DESCRIPTION
**Pull request by subash-poudel**
Finish timesheet api

### CODE REVIEW

File: src/controllers/projectTaskController.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { Router } from 'express';
+import HttpStatus from 'http-status-codes';
+import * as projectTaskService from '../services/projectTaskService';
+import logger from '../utils/logger';
+// import { findUser, userValidator } from '../validators/userValidator';
```
Comment (aviskarkc10) [inline src/controllers/projectTaskController.js:5]: .

File: src/controllers/projectTaskController.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+const router = Router();
+
+/**
+ * GET /api/projecttasks
+ */
+router.get('/', (req, res, next) => {
+  logger.info('Project Task Controller fetch all tasks.');
+  projectTaskService
```
Comment (aviskarkc10) [inline src/controllers/projectTaskController.js:14]:
```suggestion
  try {
      const projectTasks = await projectTaskService.fetchAll();
      
      res.json({  data });
  } catch (error) {
      next(error);
  }
```

File: src/controllers/projectTaskController.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { Router } from 'express';
```
Comment (aviskarkc10) [inline src/controllers/projectTaskController.js:1]: better to separate file imports according to their types
```suggestion
import { Router } from 'express';
 import HttpStatus from 'http-status-codes';
 
 import logger from '../utils/logger';
 
 import * as projectTaskService from '../services/projectTaskService';
```

File: src/controllers/timesheetController.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import { Router } from 'express';
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:1]: same here

File: src/controllers/timesheetController.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import { findTimesheet, timesheetPostValidator } from '../validators/timesheetValidator';
+import logger from '../utils/logger';
+
+const router = Router();
+
+/**
+ * GET /api/timesheet
+ */
+router.get('/', (req, res, next) => {
+  // logger.info('Controller get all');
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:13]: .

File: src/controllers/timesheetController.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  // logger.info(`from date ${req.query.from}`);
+  // logger.info(`to date ${req.query.to}`);
+  timesheetService
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:16]: using `async/await`

File: src/controllers/timesheetController.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    .catch(err => next(err));
+});
+
+/**
+ * GET /api/timesheet/:id
+ */
+router.get('/:id', findTimesheet, (req, res, next) => {
+  logger.info('Controller get by id');
+
+  timesheetService
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:28]: use `async/await`

File: src/controllers/projectTaskController.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import HttpStatus from 'http-status-codes';
+import * as projectTaskService from '../services/projectTaskService';
+import logger from '../utils/logger';
+// import { findUser, userValidator } from '../validators/userValidator';
+
+const router = Router();
+
+/**
+ * GET /api/projecttasks
```
Comment (aviskarkc10) [inline src/controllers/projectTaskController.js:10]: we probably don't need the `/api` part in the doc block. Also separate routes with either `-` or `_`. For eg: `/api/project-tasks`

File: src/controllers/timesheetController.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    .fetchTimesheet(req.params.id)
+    .then(data => res.json({ data }))
+    .catch(err => next(err));
+});
+
+/**
+ * POST /api/timesheet
+ */
+router.post('/', timesheetPostValidator, (req, res, next) => {
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:37]: how about just `timesheetValidator`

File: src/controllers/timesheetController.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  logger.info('Controller create');
+
+  timesheetService
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:40]: `async/await`

File: src/controllers/timesheetController.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    .then(data => res.status(HttpStatus.CREATED).json({ data }))
+    .catch(err => next(err));
+});
+
+/**
+ * PUT /api/timesheet/:id
+ */
+router.put('/:id', findTimesheet, timesheetPostValidator, (req, res, next) => {
+  logger.info('Timesheet put request ' + req.body);
+  timesheetService
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:51]: `async/await`

File: src/controllers/timesheetController.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    .catch(err => next(err));
+});
+
+/**
+ * DELETE /api/timesheet/:id
+ */
+router.delete('/:id', findTimesheet, (req, res, next) => {
+  logger.info('Controller delete');
+
+  timesheetService
```
Comment (aviskarkc10) [inline src/controllers/timesheetController.js:63]: .

File: src/knexfile.js [hunk 13]
```diff
@@ -1,5 +1,11 @@
 require('@babel/register');
 require('dotenv').config({ path: __dirname + '/../.env' });
+const types = require('pg').types;
+
+// 1082 is the column id of date field. the code below returns the date in the same format as
+// it is saved in database
+
+types.setTypeParser(1082, val => val);
```
Comment (aviskarkc10) [inline src/knexfile.js:8]: let's use the value from constants

File: src/models/timesheet.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+  get hasTimestamps() {
+    return true;
+  }
+
+  // parse() {
+  //   return '2018-11-11';
+  // }
+
+  // summary: function() {
```
Comment (aviskarkc10) [inline src/models/timesheet.js:20]: .

File: src/seeds/timesheet_task_seed.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import { TIMESHEET_TASKS_TABLE } from '../constants/tables';
+import logger from '../utils/logger';
+
+const taskList = [
```
Comment (aviskarkc10) [inline src/seeds/timesheet_task_seed.js:4]: put this in constants

File: src/services/timesheetService.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+/**
+ * Delete a timesheet.
+ *
+ * @param  {Number}  id
+ * @return {Promise}
+ */
+export function deleteTimesheet(id) {
+  const timesheet = new TimeSheet({ id }).fetch();
```
Comment (aviskarkc10) [inline src/services/timesheetService.js:57]: use async/await with try catch block

File: src/services/timesheetService.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Update a timesheet.
+ *
+ * @param  {Number}  id
+ * @param  {Object}         data
+ * @return {Promise}
+ */
+export function updateTimesheet(id, data) {
+  return new TimeSheet({ id }).save(data);
```
Comment (aviskarkc10) [inline src/services/timesheetService.js:47]: check if timesheet with that id exists

File: src/services/timesheetService.js [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+/**
+ * Fetch a timesheet.
+ *
+ * @param  {Number}  id
+ * @return {Promise}
+ */
+export function fetchTimesheet(id) {
+  return new TimeSheet({ id }).fetch().then(timesheet => {
```
Comment (aviskarkc10) [inline src/services/timesheetService.js:20]: use async await

File: src/validators/timesheetSchema.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
+import Joi from 'joi';
+import { TIMESHEET_STATUS_SAVED, TIMESHEET_STATUS_SUBMITTED } from '../constants/timesheetStatus';
+
+const TIMESHEET_POST_SCHEMA = {
+  //  2018-10-31
```
Comment (aviskarkc10) [inline src/validators/timesheetSchema.js:5]: .

File: src/validators/timesheetSchema.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+  //  2018-1-1
+  date: Joi.date()
+    .label('Date')
+    .required(),
```
Comment (aviskarkc10) [inline src/validators/timesheetSchema.js:9]: you can write a regex to check if the format of date matters

File: src/validators/timesheetValidator.js [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Validate timesheet existence.
+ *
+ * @param  {object}   req
+ * @param  {object}   res
+ * @param  {function} next
+ * @return {Promise}
+ */
+function findTimesheet(req, res, next) {
```
Comment (aviskarkc10) [inline src/validators/timesheetValidator.js:13]: this function should be a service

File: src/validators/timesheetValidator.js [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+  return timesheetService
+    .fetchTimesheet(req.params.id)
```
Comment (aviskarkc10) [inline src/validators/timesheetValidator.js:15]: async/await

File: src/validators/timesheetValidator.js [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Validate create/update timesheet request.
+ *
+ * @param  {object}   req
+ * @param  {object}   res
+ * @param  {function} next
+ * @return {Promise}
+ */
+function timesheetPostValidator(req, res, next) {
+  return validate(req.body, TIMESHEET_POST_SCHEMA)
```
Comment (aviskarkc10) [inline src/validators/timesheetValidator.js:29]: async/await

File: src/controllers/projectTaskController.js (hunk 1 above)
Comment (subash-poudel) [reply src/controllers/projectTaskController.js:5]: 🆗

File: src/controllers/projectTaskController.js (hunk 2 above)
Comment (subash-poudel) [reply src/controllers/projectTaskController.js:14]: 🆗

File: src/controllers/projectTaskController.js (hunk 3 above)
Comment (subash-poudel) [reply src/controllers/projectTaskController.js:1]: 🆗

File: src/controllers/timesheetController.js (hunk 4 above)
Comment (subash-poudel) [reply src/controllers/timesheetController.js:1]: 🆗

File: src/controllers/timesheetController.js (hunk 5 above)
Comment (subash-poudel) [reply src/controllers/timesheetController.js:13]: 🆗

**Pull request merged**